File: various/src/main/java/com/chrnie/various/Various.java
package com.chrnie.various;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import com.chrnie.various.ViewHolder.Factory;
import java.util.ArrayList;
import java.util.List;

public final class Various {

  private Various() {
    throw new IllegalStateException("none constructor");
  }

  @NonNull
  public static Builder of(@NonNull List<?> dataList) {
    return Various.of(dataList, DefaultItemMatcherFactory.getInstance());
  }

  @NonNull
  public static Builder of(@NonNull List<?> dataList, @NonNull ItemMatcher.Factory factory) {
    return new Builder(dataList, factory);
  }

  public interface OnCreateCallback<VH extends RecyclerView.ViewHolder> {

    @NonNull
    VH onCreate(@NonNull LayoutInflater inflater, @NonNull ViewGroup container);
  }

  public interface OnBindCallback<VH extends RecyclerView.ViewHolder, T> {

    void onBind(@NonNull VH holder, @NonNull T data, @NonNull List<Object> payloads);
  }

  public static class Builder {

    final List<?> dataList;
    final ItemMatcher.Factory factory;
    final List<Item> itemList = new ArrayList<>(2);

    Builder(List<?> dataList, ItemMatcher.Factory factory) {
      this.dataList = dataList;
      this.factory = factory;
    }

    @NonNull
    public <VH extends RecyclerView.ViewHolder, T> Builder register(
        @NonNull Class<T> dateType,
        @NonNull OnCreateCallback<VH> onCreateCallback) {
      return register(dateType, onCreateCallback, (holder, data, payloads) -> {
        // empty;
      });
    }

    @NonNull
    public <VH extends RecyclerView.ViewHolder, T> Builder register(
        @NonNull Class<T> dateType,
        @NonNull OnCreateCallback<VH> onCreateCallback,
        @NonNull OnBindCallback<VH, ? super T> onBindCallback) {
      itemList.add(new Item(dateType, onCreateCallback, onBindCallback));
      return this;
    }

    @NonNull
    public <T> Builder register(
        @NonNull Class<T> dateType,
        @NonNull Factory<? super T, ? extends ViewHolder<? super T>> factory) {
      return register(dateType, factory::create, ViewHolder::bind);
    }

    @NonNull
    public RecyclerView.Adapter<RecyclerView.ViewHolder> build() {
      return new AdapterImpl(this);
    }
  }

  public interface ViewHolderCallback {

    default boolean onFailedToRecycleView() {
      return false;
    }

    default void onViewAttachedToWindow() {
    }

    default void onViewDetachedFromWindow() {
    }

    default void onViewRecycled() {
    }
  }
}

File: various/src/main/java/com/chrnie/various/Various.kt
package com.chrnie.various

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import kotlin.reflect.KClass

object Various {

    class Builder<DATA : Any> @JvmOverloads constructor(private val itemMatcherFactory: ItemMatcher.Factory = DefaultItemMatcherFactory) {
        private val itemList = ArrayList<Item<*, *>>()

        fun <T : DATA, VH : RecyclerView.ViewHolder> register(
                dataType: KClass<in T>,
                viewHolderBinder: ViewHolderBinder<T, VH>
        ): Builder<DATA> {
            return register(dataType.java, viewHolderBinder)
        }

        fun <T : DATA, VH : RecyclerView.ViewHolder> register(
                dataType: Class<in T>,
                viewHolderBinder: ViewHolderBinder<T, VH>
        ): Builder<DATA> {
            itemList.add(Item(dataType, viewHolderBinder))
            return this
        }

        @JvmOverloads
        fun <T : DATA, VH : RecyclerView.ViewHolder> register(
                dataType: KClass<in T>,
                onCreateViewHolderCallback: OnCreateViewHolderCallback<VH>,
                onBindViewHolderCallback: OnBindViewHolderCallback<T, VH> = { _, _, _ -> }
        ): Builder<DATA> {
            return register(dataType.java, onCreateViewHolderCallback, onBindViewHolderCallback)
        }

        @JvmOverloads
        fun <T : DATA, VH : RecyclerView.ViewHolder> register(
                dataType: Class<in T>,
                onCreateViewHolderCallback: OnCreateViewHolderCallback<VH>,
                onBindViewHolderCallback: OnBindViewHolderCallback<T, VH> = { _, _, _ -> }
        ): Builder<DATA> {
            val binder = LambdaViewHolderBinder(onCreateViewHolderCallback, onBindViewHolderCallback)
            val item = Item(dataType, binder)
            itemList.add(item)
            return this
        }

        fun build(): Various.Adapter<DATA> {
            val itemMatcher = itemMatcherFactory.create(ArrayList(itemList))
            return Adapter(itemMatcher)
        }
    }

    class Adapter<DATA : Any> internal constructor(private val itemMatcher: ItemMatcher) :
            RecyclerView.Adapter<RecyclerView.ViewHolder>() {

        var dataList: List<DATA> = emptyList()

        override fun getItemCount(): Int = dataList.size

        override fun getItemViewType(position: Int): Int {
            val data = dataList[position]
            return itemMatcher.requestViewType(data)
        }

        override fun getItemId(position: Int): Long {
            val viewType = getItemViewType(position)
            val binder = itemMatcher.requestViewHolderBinder(viewType)
            val data = dataList[position]
            val id = binder.getItemId(data)
            return viewType.toLong() shl 32 or id.toLong()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val binder = itemMatcher.requestViewHolderBinder(viewType)
            return binder.onCreateViewHolder(LayoutInflater.from(parent.context), parent)
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            val data = dataList[position]
            val binder = itemMatcher.requestViewHolderBinder(holder.itemViewType)
            binder.onBindViewHolder(holder, data, null)
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int, payloads: List<Any>?) {
            val data = dataList[position]
            val binder = itemMatcher.requestViewHolderBinder(holder.itemViewType)
            binder.onBindViewHolder(holder, data, payloads)
        }

        override fun onViewAttachedToWindow(holder: RecyclerView.ViewHolder) {
            val binder = itemMatcher.requestViewHolderBinder(holder.itemViewType)
            binder.onViewAttachedToWindow(holder)
        }

        override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
            val binder = itemMatcher.requestViewHolderBinder(holder.itemViewType)
            binder.onViewDetachedFromWindow(holder)
        }

        override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
            val binder = itemMatcher.requestViewHolderBinder(holder.itemViewType)
            binder.onViewRecycled(holder)
        }

        override fun onFailedToRecycleView(holder: RecyclerView.ViewHolder): Boolean {
            val binder = itemMatcher.requestViewHolderBinder(holder.itemViewType)
            return binder.onFailedToRecycleView(holder)
        }
    }
}